### DISCUSSION TITLE
When to increase __version__ in project's __init__.py?

### DISCUSSION THREAD

**Discussion by mathieuboudreau** [neuropoly/axondeepseg#118] [Bug reports, suggestions for code refactoring]
Currently on master it's listed as v0.5:

https://github.com/neuropoly/axondeepseg/blob/4fb50073a22d6358451567a4004236796d84058c/AxonDeepSeg/__init__.py#L1

Even though the most recently tagged release is 0.4.1:

https://github.com/neuropoly/axondeepseg/tags

(Also, notice the most latest version on ADS PyPi is actually only 0.3: https://pypi.org/project/AxonDeepSeg/#history)

I'm assuming that @alzaia increased the version in __init__.py to 0.5 after tagging 0.4.1, because he assumed 0.5 would be the next release (it appears that way from his instructions [on the wiki](https://github.com/neuropoly/axondeepseg/wiki/How-to-create-a-new-release) ? But I actually want to tag my Sentry PR to be release 0.4.2 since it's only a minor change (and only minor changes have been done since 0.4.1). I feel that it would be more natural to only change the value in __init__.py when the version actually gets tagged, since we don't know which iteration of the version is actually coming up next (*.*.x, *.x.*, or x.*.*), à la [ "semantic versioning"](https://semver.org). 

If I tag my PR to 0.4.2, then there will be several commits prior to it that will have a version above it (0.5), which might lead to some confusions.  Thoughts?

**Comment by perone** [comment]
If we use and recommend users to install using only:

`pip install axondeepseg` or `python setup.py install`

We can do like that:

1. Soon after releasing the `0.1` version for instance, we just change the `__init__.py` version (_and this variable should be used by `setup.py` as well for the `conf.py` documentation, so we just have to change one place_) to be `0.2dev` (Version: 0.2, development pre-release);
2. We don't tag this new version yet, tagging will only happen at the moment of the release to official PyPI, so when there is a tag on the repository, we know that this tag is the exact version of the matching PyPI release;
3. If we need to release a version for a user to test, we can issue a pre-release such as `0.2dev`, so user can explicitly install it with `pip install axondeepseg=0.2dev` or `pip install --pre axondeepseg`. Since this is pre-release, if any user tries to install with `pip install axondeepseg`, it will install the latest stable version `0.1`, it's a way to issue pre-releases without impacting everyone;
4. Soon after we finished the `0.2` release, and want to do the final release, we just change the version from `0.2dev` to `0.2` and then issue the release and tag it;

I would prefer to keep just the `[major].[minor]` version to keep it simple, where changes in the major are usually breaking changes or new features and changes in minor are bug fixes or something minor.

**Comment by mathieuboudreau** [comment]
That sounds like a great strategy – thanks @perone !

> I would prefer to keep just the [major].[minor] version to keep it simple, where changes in the major are usually breaking changes or new features and changes in minor are bug fixes or something minor.

Since currently we've been treating \*.x.\* as major and \*.\*.x as minor, should we maybe move to version 1.0 (the paper's been published anyways) after the upcoming Sentry and Testing Suite PR's are merged, and start your `[major].[minor]` strategy from there?

**Comment by perone** [comment]
I also think it's a good idea to move to the 1.0 version.

**Comment by mathieuboudreau** [comment]
@perone I tried creating a version using your nomenclature ("0.5dev" in __ini__.py) and packaging it for PyPi using [these instructions](https://github.com/neuropoly/axondeepseg/wiki/How-to-update-axondeepseg-wheel-in-PyPI).

It's showing up on the website (https://pypi.org/project/AxonDeepSeg/0.5.dev0/), but when trying to install it, my local pip can't find that dev version: 

```

(ads_venv) mathieuboudreau@brassens:~$ pip install --pre AxonDeepSeg==0.5.dev0
Collecting AxonDeepSeg==0.5.dev0
  Could not find a version that satisfies the requirement AxonDeepSeg==0.5.dev0 (from versions: 0.1, 0.2, 0.4, 0.5, 0.33)
No matching distribution found for AxonDeepSeg==0.5.dev0
(ads_venv) mathieuboudreau@brassens:~$ 

```

Do you know if it take a while before pip is updated to be able to find new releases? (I've been trying for a couple of hours). I tried downloading and installing the wheel file manually, and that worked, I just can't get it using the pip command.



(Also, I'm not sure why it can "find" version versions 0.2, 0.33, and 0.4, considering only 0.1 and 0.3 should be on there: https://pypi.org/project/AxonDeepSeg/#history. I'm assuming that if you skip a version, it fills in the missing integers versions with the same one? And 0.33 is simply 0.3 with a "3" added after it for the alpha release?)

**Comment by mathieuboudreau** [comment]
Nevermind, it finally showed up after 4 hours or so. 🎉
